feat: add tutorial and collection cards

zchsh opened the pull request:
## ✅ "Ready for Review" checklist

<!--
Check these items off in the GitHub PR UI as you complete them. If one of these items is not relevant to your PR, e.g. you do not have Asana ticket to go with them, you can remove them from the list.
-->

- [x] The Vercel preview link has been added to this PR's description
- [x] The Asana task has been added to this PR's description
- [x] This PR's link has been added to the "📐 Pull Request" field on the Asana task
- [x] This Vercel preview link has been added to the "🔍 Preview" field on the Asana task
- [x] The description template has been filled in below

---

## 🔗 Relevant links

- [Preview - `CollectionCard`][swingset-collection-card] 🔎
- [Preview - `TutorialCard`][swingset-tutorial-card] 🔎
- [Asana task](https://app.asana.com/0/1201987349274776/1201654639085737/f) 🎟️

## 🗒️ What

This PR implements `components/tutorial-card` & `components/collection-card`. It also replaces any existing stubbed implementations of these components with the new implementations.

## 🤷 Why

To finish building out the tutorial views.

## 🛠️ How

- Sets up a `components/tutorial-collection-cards` folder for shared elements used in both cards
- Composes those shared elements to build `components/tutorial-card`
- Composes those shared elements to build `components/collection-card`

## 📸 Screenshots

<details>

<summary>Tutorial Card</summary>

Design:

<img width="985" alt="CleanShot 2022-04-21 at 15 59 02@2x" src="https://user-images.githubusercontent.com/4624598/164542793-290af1db-8267-401b-9405-2893448400ab.png">

Implementation:

<img width="792" alt="tutorial-card" src="https://user-images.githubusercontent.com/4624598/164541756-9d105fa5-16a5-4571-9c74-d81feecc4991.png">

</details>

<details>

<summary>Collection Card</summary>

Design:

<img width="962" alt="CleanShot 2022-04-21 at 15 59 44@2x" src="https://user-images.githubusercontent.com/4624598/164542906-d4cca7bf-e38d-44de-8888-30f51f9bf901.png">

Implementation: 

<img width="788" alt="collection-card" src="https://user-images.githubusercontent.com/4624598/164541829-09f03d94-790d-4660-87a7-a664da6b6ab7.png">

</details>

## 🧪 Testing

- Visit Swingset, and review the [`Collection Card`][swingset-collection-card] & [`Tutorial Card`][swingset-tutorial-card] pages, and make sure things look correct
    - Note that the `description` on both types of cards is truncated to 3 lines using CSS.
    - [x] Truncation should function as expected in all major browsers (tested in Safari, Firefox, & Chrome)
- Visit a tutorial view that shows "Featured Collections", such as [the Waypoint install tutorial][waypoint-install-tutorial]
    - [ ] Ensure the "Also appears in" cards at the bottom of the page look correct.
- Visit [the home page][home-page]
    - [ ] Ensure the tutorial cards partway down the page look as expected.
    - Note: the data for these cards is stubbed in placeholder. So it is unlikely to look correct.

> Note: If you want, you can review the `CollectionCard` & `TutorialCard` components [in Storybook][storybook]. I found mobile behaviour & truncation a little easier to test.

## 💭 Anything else?

- This PR does _not_ try to use the new components in new places. It only replaces existing stubbed implementations. These components will be used on our tutorials views in subsequent PRs (which will large be constructing grid layouts with these cards).
- Super minor nitpick: The `badge` size is just _slightly_ off. Fixing this would require changes to the `badge` component, and I think would implicate the `badge` component's current `display: inline-flex`. This is very much a pixel-pushing issue, like 2px of extra whitespace, so _maybe_ will try to tweak this before June 20, but definitely need to understand the best way forward to do so.

[home-page]: https://dev-portal-git-zsadd-tutorial-collection-cards-hashicorp.vercel.app
[waypoint-install-tutorial]: https://dev-portal-git-zsadd-tutorial-collection-cards-hashicorp.vercel.app/waypoint/tutorials/get-started-docker/get-started-install
[swingset-collection-card]: https://dev-portal-git-zsadd-tutorial-collection-cards-hashicorp.vercel.app/swingset/components/collectioncard
[swingset-tutorial-card]: https://dev-portal-git-zsadd-tutorial-collection-cards-hashicorp.vercel.app/swingset/components/tutorialcard
[storybook]: https://dev-portal-with-storybook.vercel.app/?path=/story/components-tutorialcard--playground

zchsh commented:
@kendallstrautman This PR implements Tutorial and Collection cards.

I've taken the approach of implementing composable elements in `components/tutorial-collection-cards`, and then using those elements in `components/tutorial-card` & `components/collection-card`. This results in more files & more lines of code than might be possible than with more of "everything in one file", but hopefully with the benefit of flexibility & maintainability going forward.

Let me know what you think, happy to adjust the approach here as needed 🙇‍♂️

kendallstrautman commented on src/components/collection-card/index.tsx:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Render a card that links to a collection.
+ *
+ * TODO: note that not all collection links will work, for now.
```
This should be finished, it was handled via #267 👌 So long as we use the `getCollectionSlug` helper function, it will be handled properly.

zchsh replied:
oh cool nice! 🤩 🎉  I'll drop this comment, had just kept it cause it was in the previous stubbed collection card in `tutorial-view`

kendallstrautman commented on src/components/tutorial-card/types.ts:
```diff
@@ -0,0 +1,38 @@
+  /**
+   * A list of product slugs, representing products used in the tutorial.
+   */
+  productsUsed: ProductSlug[]
```
```suggestion
  productsUsed: ProductOption[]
```
There's a subtle difference but since the data is coming from the learn api I think we should use this.

kendallstrautman replied:
Thats in `lib/learn-client/types` I think

zchsh replied:
✅ Updated to use `ProductOption`

kendallstrautman commented on src/components/tutorial-collection-cards/components/card-badges/index.tsx:
```diff
@@ -0,0 +1,64 @@
+  vault: <ProductIcon productSlug="vault" />,
+  vagrant: <ProductIcon productSlug="vagrant" />,
+  waypoint: <ProductIcon productSlug="waypoint" />,
+  hcp: <ProductIcon productSlug="hcp" />,
```
In learn we call this `cloud`, just to be clear as what's coming in the data, which references `ProductOption`. Also sentinel isn't a valid learn product so I don't think we need to include that. 
```suggestion
  cloud: <ProductIcon productSlug="hcp" />,
```

kendallstrautman replied:
💭 Also the `ProductOption` from learn client types is an enum so it could help you generate this more easily?

zchsh replied:
Ah wait so I realize here that i need to _drop_ `cloud` or `hcp` right? Since `cloud` isn't a possible option? (Much like `sentinel`?)

zchsh replied:
✅ I've updated to remove the `hcp` key (so no `cloud` key either). Lemme know if I should do something different 🙇‍♂️

kendallstrautman commented on src/components/tutorial-collection-cards/components/card-badges/index.tsx:
```diff
@@ -0,0 +1,64 @@
+    <ul className={s.root}>
+      {badges.map((badge: CardBadgeOption) => {
+        // Handle null Sentinel icon case
+        if (badge == 'sentinel') {
```
We could probably remove this since it's not a possible learn product atm

zchsh replied:
✅ Awesome, much cleaner - removed!

kendallstrautman commented on src/components/tutorial-collection-cards/components/card-badges/types.ts:
```diff
@@ -0,0 +1,7 @@
+import type { ProductSlug } from 'types/products'
+
+export type CardBadgeOption = ProductSlug | 'video' | 'interactive'
```
```suggestion
export type CardBadgeOption = ProductOption | 'video' | 'interactive'
```

kendallstrautman commented on src/components/tutorial-collection-cards/components/card-body/index.tsx:
```diff
@@ -0,0 +1,17 @@
+function CardBody({ text }: { text: string }) {
+  return (
+    <p className={s.body}>
+      <TruncateMaxLines
```
Slick ✨

kendallstrautman commented on src/components/tutorial-collection-cards/components/card-heading/card-heading.module.css:
```diff
@@ -0,0 +1,28 @@
+
+.logo {
+  display: block;
+  height: 20px;
```
Some of the logos look a little off, small. But we probably just need to adjust the whitespace maybe?

zchsh replied:
Great call - just flagged this with Brayden. I think since our `Terraform Cloud` logo is coming from a blessed repo, while all the others are local `.png`s, the easiest path forward is to make em all look the same size with some top & bottom whitespace built into said PNGs.

I'll plan to patch this up in a follow-up PR, after the collection views are done... or during VQA, I would be surprised if it doesn't come up again at that point if we do forget to do it (feels like it's lower priority, but lemme know if it feels more critical to address 👍).

kendallstrautman replied:
Perfect, definitely lower priority for sure.

kendallstrautman commented on src/components/tutorial-collection-cards/components/card-heading/company-logo/index.tsx:
```diff
@@ -0,0 +1,28 @@
+      alt=""
+      /**
+       * Note: for use in CollectionCard, consumer should provide meaningful
+       * text using the card heading element.
```
Good note 👌

kendallstrautman commented on src/components/tutorial-collection-cards/components/card-heading/index.tsx:
```diff
@@ -0,0 +1,24 @@
+      <span className={classNames(hasLogo && 'g-screen-reader-only')}>
+        {text}
+      </span>
+      {hasLogo && (
```
If this is shared with the tutorial card, perhaps we write a note that only collection cards should render a logo?

zchsh replied:
Great call... Looking at this again, I think I should probably move the `logo` stuff into `components/collection-card`. Going to do that now - I'll have to add more CSS for logo sizing tweaks (chatted with Brayden via Slack on that), so don't want collection-card-specific stuff to overwhelm these shared elements 👍

zchsh replied:
✅ Done. Moved `logo` stuff into `collection-card`, was a bit of a refactor, but I think it's much clearer now.

kendallstrautman commented on src/components/tutorial-collection-cards/components/index.tsx:
```diff
@@ -0,0 +1,4 @@
+export * from './card-heading'
+export * from './card-body'
+export * from './card-eyebrow'
+export * from './card-badges'
```
🚀

kendallstrautman commented on src/components/tutorial-collection-cards/components/truncate-max-lines/index.tsx:
```diff
@@ -0,0 +1,24 @@
+import s from './truncate-max-lines.module.css'
+import { TruncateMaxLinesProps } from './types'
+
+function TruncateMaxLines({
```
💖

kendallstrautman commented on src/views/homepage/components/learn-section/index.tsx:
```diff
@@ -55,7 +55,7 @@ export default function LearnSection({
           }
         >
-          {tutorials.map((tutorial) => {
+          {tutorials.map((tutorial: TutorialCardProps) => {
```
Where is this tutorials data coming from 🤔 ?

kendallstrautman replied:
Oh I see it's hardcoded. nvm

zchsh replied:
Hard-coded in the homepage JSX! I figured for this PR we're not gonna change that.

kendallstrautman replied:
For sure! Just a reminder for me that I need to fix it ha

zchsh replied:
Oops, crossed wires there - didn't see you post that yet 😅 ⚡

kendallstrautman commented on src/components/tutorial-card/types.ts:
```diff
@@ -0,0 +1,38 @@
+  /**
+   * A string representation the duration of the tutorial, such as "10mins".
+   */
+  duration: string
```
So you see the duration formatting calculation happening higher in the tree? Or would it feel reasonable for the tutorial card to do this?

zchsh replied:
Ah i assumed it'd happen higher in the tree! Definitely feels reasonable for tutorial card to do this.

I think I'd plan to change `duration: string` here to `durationMinutes: number`:
1. Does that sound approximately right?
2. Would you be okay if we defer this change to when I'm implementing the collection view (slash rendering tutorial cards with some real data?). Just thinking it'll be easier to validate that things are working as expected if I'm using it with "wired up" data.

kendallstrautman replied:
Ya totally, it's v minor 🙌

zchsh replied:
Sweet, I'll defer it till the next PR 👍

kendallstrautman commented on src/views/tutorial-view/components/featured-in-collections/helpers.ts:
```diff
@@ -1,15 +1,39 @@
+ *
+ * Should we parse the tutorials within a collection in order
+ * to show a more complete list? Or will this result in more of a
+ * "product icon overload" on these cards?
```
🤔 Hmmmmm..... Currently on learn the collection card logos render their theme only.

kendallstrautman replied:
My hunch is to stick with current behavior, if we want to adjust cloud logos to also render all their tutorials product options, we can adjust later? 🤷‍♀️

kendallstrautman replied:
If we do that, we may want to rethink using the `productsUsed` prop, or just writing a note that this is actually just `theme`

zchsh replied:
My aim with the implementation here is to match that behaviour! Added this note cause i'm seeing many product icons in the designs:

<img width="717" alt="CleanShot 2022-04-21 at 17 15 44@2x" src="https://user-images.githubusercontent.com/4624598/164553667-7eca29b3-cefd-415f-9e8f-8cc02e74dc21.png">

I think this is currently impossible? So just added this as a reminder that we should check in with design on behaviour here, at some point. Let me know if that make sense - if so I'll open & link to an Asana ticket in the comment, if not I'll drop the comment 👍

kendallstrautman replied:
ya that sounds good 😄

braydenlove replied:
Sorry I was just testing how many badges we could fit at each size here. I know there can only be one currently but I wanted to be flexible for the future as well!

zchsh replied:
@braydenlove sweet, makes total sense!

kendallstrautman commented on src/views/tutorial-view/components/featured-in-collections/helpers.ts:
```diff
@@ -1,15 +1,39 @@
+    // which is compatible with dev-dot's ProductSlug types
+    return [theme]
+  } else {
+    // For other themes, return "hcp", for a generic "H" logo
```
Currently in learn if a collection has a non product theme we don't render a logo at all...

kendallstrautman replied:
See the cloud page for an example - https://learn.hashicorp.com/cloud

We can totally render the generic cloud logo, just pointing it out

zchsh replied:
Ah cool, I'll drop this then 👍

zchsh replied:
ah sorry another case of crossed wires - do you think i should add back returning `cloud` here?

kendallstrautman replied:
I think we should just render no logo unless its a valid `product`

zchsh replied:
✅ Sweet, done 👍 Can always adjust later, of course 🎛️

kendallstrautman reviewed (changes requested):
Looking great! 🚤  I left some minor suggestions relating to the product slug / option types. Once we align on that I'd say this is good to go!

zchsh commented:
@kendallstrautman Thanks for the quick review, I'll work on addressing feedback here and try do that fast enough that we can merge today 😅 Excited to start using these on the remaining tutorial views 🌞

zchsh commented:
@kendallstrautman Did my best to address your feedback! Updates since last review:

- Switch to `ProductOption` type for consistency
- Dropped `sentinel` & `hcp` / `cloud` logo handling to match existing behaviour (clearer now thanks to `ProductOption`)
- Moved `CompanyLogo` stuff into `components/collection-card/components`, feels much more "correct"
- Patched up the logo relative sizes issue
    - Slacked with Brayden and figure I might as well get this done ASAP
    - This felt easier to do now that `CompanyLogo` stuff had a less generic home (it's clearer we're just making these adjustments for `collection-card`, for now)
    - Was also just curious to experiment with approaches here as "scale logos with no padding to be optically consistent sizes" has been something that's haunted me since like Day 1. I don't think I have something _fully_ abstract at all, but thanks to `next/image`'s scaling magic it doesn't feel too far off.

Here's a preview of the optically corrected logo sizes
(cc @braydenlove, I think we're aiming to merge this & keep moving, but definitely add any notes on further scaling correction we could do & I can add em to our backlog! 💖 )

<details>

<summary>Before</summary>

<img width="761" alt="collection-card" src="https://user-images.githubusercontent.com/4624598/164563743-2d0a58c1-d034-4688-9a40-a1884ab51d8c.png">

</details>

<details>

<summary>After</summary>

<img width="1073" alt="CleanShot 2022-04-21 at 18 39 38@2x" src="https://user-images.githubusercontent.com/4624598/164563702-f9e62323-d1d3-4254-b2bf-5b19d008080f.png">

</details>

Let me know what you think 🙇‍♂️

zchsh commented:
(rebuilding to sync with `main`. existing deploy previews are otherwise up-to-date)

braydenlove commented:
Awesome work @zchsh + @kendallstrautman , these look awesome. SHIP IT

kendallstrautman reviewed (approved):
Looks great!!! :shipit:
